File: apps/macos/DNSPilotMac/Sources/DNSPilotMacCore/ApplyPlanRunner.swift
import Foundation

public enum ApplyPlanRunnerError: Error, Equatable {
    case processFailed(String)
}

public enum ApplyPlanConfidence: String, Equatable, Sendable {
    case high
    case medium
    case low
    case inconclusive
}

public enum ApplyPlanGateHealth: String, Equatable, Sendable {
    case healthy
    case degraded
    case failed
    case inconclusive
}

public struct ApplyPlanRequest: Equatable, Sendable {
    public let platformID: String
    public let profileDatabaseURL: URL?
    public let profileID: String?
    public let testedResolver: String?
    public let confidence: ApplyPlanConfidence
    public let gateHealth: ApplyPlanGateHealth
    public let vpnActive: Bool
    public let mdmProfileActive: Bool
    public let corporateDNSDetected: Bool
    public let captivePortalDetected: Bool

    public init(
        platformID: String = "macos-store",
        profileDatabaseURL: URL? = nil,
        profileID: String?,
        testedResolver: String? = nil,
        confidence: ApplyPlanConfidence = .high,
        gateHealth: ApplyPlanGateHealth = .healthy,
        vpnActive: Bool = false,
        mdmProfileActive: Bool = false,
        corporateDNSDetected: Bool = false,
        captivePortalDetected: Bool = false
    ) {
        self.platformID = platformID
        self.profileDatabaseURL = profileDatabaseURL
        self.profileID = profileID
        self.testedResolver = testedResolver
        self.confidence = confidence
        self.gateHealth = gateHealth
        self.vpnActive = vpnActive
        self.mdmProfileActive = mdmProfileActive
        self.corporateDNSDetected = corporateDNSDetected
        self.captivePortalDetected = captivePortalDetected
    }

    public var commandArguments: [String] {
        var arguments = [
            "apply-plan",
            platformID,
            "--confidence", confidence.rawValue,
            "--gate-health", gateHealth.rawValue,
        ]
        if let profileDatabaseURL {
            arguments += ["--profile-db", profileDatabaseURL.path]
        }
        if let profileID {
            arguments += ["--profile-id", profileID]
        }
        if let testedResolver {
            arguments += ["--tested-resolver", testedResolver]
        }
        if vpnActive {
            arguments.append("--vpn-active")
        }
        if mdmProfileActive {
            arguments.append("--mdm-profile-active")
        }
        if corporateDNSDetected {
            arguments.append("--corporate-dns-detected")
        }
        if captivePortalDetected {
            arguments.append("--captive-portal-detected")
        }
        return arguments
    }
}

public struct ApplyPlanRunner {
    private let executableURL: URL
    private let processRunner: any BenchmarkProcessRunning
    private let decoder: ApplyPlanJSONDecoder

    public init(
        executableURL: URL,
        processRunner: any BenchmarkProcessRunning = FoundationBenchmarkProcessRunner(),
        decoder: ApplyPlanJSONDecoder = ApplyPlanJSONDecoder()
    ) {
        self.executableURL = executableURL
        self.processRunner = processRunner
        self.decoder = decoder
    }

    public func load(request: ApplyPlanRequest) throws -> ApplyPlan {
        let output = try processRunner.run(
            executableURL: executableURL,
            arguments: request.commandArguments,
            cancellation: nil
        )
        guard output.exitCode == 0 else {
            throw ApplyPlanRunnerError.processFailed(Self.failureMessage(from: output))
        }
        return try decoder.decode(Data(output.standardOutput.utf8))
    }

    private static func failureMessage(from output: BenchmarkProcessOutput) -> String {
        let standardError = output.standardError.trimmingCharacters(in: .whitespacesAndNewlines)
        if !standardError.isEmpty {
            return standardError
        }

        let standardOutput = output.standardOutput.trimmingCharacters(in: .whitespacesAndNewlines)
        if !standardOutput.isEmpty {
            return standardOutput
        }

        return "Apply plan command exited with code \(output.exitCode)."
    }
}

File: apps/windows/DNSPilotWindows/src/DNSPilotWindows.Core/ApplyPlanRunner.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace DNSPilotWindows.Core;

public enum ApplyPlanConfidence
{
    High,
    Medium,
    Low,
    Inconclusive,
}

public enum ApplyPlanGateHealth
{
    Healthy,
    Degraded,
    Failed,
    Inconclusive,
}

public sealed record ApplyPlanRequest(
    string? profileId,
    string? testedResolver,
    ApplyPlanConfidence confidence,
    ApplyPlanGateHealth gateHealth,
    bool vpnActive = false,
    bool mdmProfileActive = false,
    bool corporateDnsDetected = false,
    bool captivePortalDetected = false,
    string platformId = BenchmarkPlanViewModel.WindowsStorePlatformId)
{
    public IReadOnlyList<string> CommandArguments
    {
        get
        {
            var arguments = new List<string>
            {
                "apply-plan",
                platformId,
                "--confidence",
                confidence.CliValue(),
                "--gate-health",
                gateHealth.CliValue(),
            };

            if (profileId is not null)
            {
                arguments.Add("--profile-id");
                arguments.Add(profileId);
            }

            if (testedResolver is not null)
            {
                arguments.Add("--tested-resolver");
                arguments.Add(testedResolver);
            }

            if (vpnActive)
            {
                arguments.Add("--vpn-active");
            }

            if (mdmProfileActive)
            {
                arguments.Add("--mdm-profile-active");
            }

            if (corporateDnsDetected)
            {
                arguments.Add("--corporate-dns-detected");
            }

            if (captivePortalDetected)
            {
                arguments.Add("--captive-portal-detected");
            }

            return arguments;
        }
    }
}

public static class ApplyPlanJsonDecoder
{
    public static ApplyPlan Decode(string json)
    {
        var payload = JsonSerializer.Deserialize<ApplyPlanPayload>(json, JsonOptions.Default)
            ?? throw new InvalidOperationException("Apply-plan payload is empty.");
        ShellPayloadSchema.Validate(payload.SchemaVersion);

        return new ApplyPlan(
            DecisionFor(payload.Disposition),
            payload.ProfileName ?? "DNS profile",
            payload.DnsServers,
            payload.TestedResolver,
            string.Join(Environment.NewLine, payload.Notes))
        {
            PlatformId = payload.Platform,
            ApplyCapability = payload.ApplyCapability,
            Disposition = payload.Disposition,
            ProfileId = payload.ProfileId,
            CanApply = payload.CanApply,
            Notes = payload.Notes,
        };
    }

    private static ApplyDecision DecisionFor(string disposition)
    {
        return disposition switch
        {
            "guide-only" => ApplyDecision.Guide,
            "apply-with-user-approval" => ApplyDecision.Guide,
            "protect-current-dns" => ApplyDecision.Protect,
            "not-recommended" => ApplyDecision.Protect,
            "unsupported" => ApplyDecision.Block,
            _ => throw new InvalidOperationException($"Unknown apply-plan disposition '{disposition}'."),
        };
    }

    private sealed record ApplyPlanPayload(
        [property: JsonPropertyName("schema_version")] int SchemaVersion,
        [property: JsonPropertyName("platform")] string Platform,
        [property: JsonPropertyName("apply_capability")] string ApplyCapability,
        [property: JsonPropertyName("disposition")] string Disposition,
        [property: JsonPropertyName("profile_id")] string? ProfileId,
        [property: JsonPropertyName("profile_name")] string? ProfileName,
        [property: JsonPropertyName("tested_resolver")] string? TestedResolver,
        [property: JsonPropertyName("dns_servers")] IReadOnlyList<string> DnsServers,
        [property: JsonPropertyName("can_apply")] bool CanApply,
        [property: JsonPropertyName("notes")] IReadOnlyList<string> Notes);
}

public sealed class ApplyPlanRunner
{
    private readonly string _executablePath;
    private readonly ICliProcessRunner _processRunner;

    public ApplyPlanRunner(string executablePath, ICliProcessRunner? processRunner = null)
    {
        _executablePath = executablePath;
        _processRunner = processRunner ?? new SystemCliProcessRunner();
    }

    public ApplyPlan Load(ApplyPlanRequest request)
    {
        var output = _processRunner.Run(_executablePath, request.CommandArguments, progressHandler: null);
        CliContractRunnerErrors.EnsureSuccess("Apply-plan", output);
        return ApplyPlanJsonDecoder.Decode(output.StandardOutput);
    }
}

internal static class ApplyPlanEnumExtensions
{
    public static string CliValue(this ApplyPlanConfidence confidence)
    {
        return confidence switch
        {
            ApplyPlanConfidence.Medium => "medium",
            ApplyPlanConfidence.Low => "low",
            ApplyPlanConfidence.Inconclusive => "inconclusive",
            _ => "high",
        };
    }

    public static string CliValue(this ApplyPlanGateHealth health)
    {
        return health switch
        {
            ApplyPlanGateHealth.Degraded => "degraded",
            ApplyPlanGateHealth.Failed => "failed",
            ApplyPlanGateHealth.Inconclusive => "inconclusive",
            _ => "healthy",
        };
    }
}
